scripts/maintenance/upgrade-NULL

```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/awk -f
+#
+## Copyright (C) 1996-2023 The Squid Software Foundation and contributors
+##
+## Squid software is distributed under GPLv2+ license and includes
+## contributions from numerous individuals and organizations.
+## Please see the COPYING and CONTRIBUTORS files for details.
+##
+
+/NULL/ {
+  # NULL is deprecated in C++11 by nullptr.
+  if ( FILENAME ~ /(\.cci?|src\/|tools\/)$/ ) {
+    gsub(/return[ ]+\(?NULL\)?;/, "return nullptr;");
+    gsub(/RETVAL\([ ]?NULL[ ]?\)/, "RETVAL(nullptr)");
+    gsub(/[ ]=[ ]?NULL[;]/, " = nullptr;");
+    gsub(/[ ]=[ ]?NULL[,]/, " = nullptr,");
+    gsub(/[ ]=[ ]?NULL[)]/, " = nullptr)");
+    gsub(/\([ ]?NULL[ ]?,/, "(nullptr,");
+    gsub(/{[ ]?NULL[ ]?,/, "{nullptr,");
+    gsub(/,[ ]?NULL[ ]?,/, ", nullptr,");
+    gsub("  NULL,", "  nullptr,");
+    gsub(/,[ ]?NULL[ ]?}/, ", nullptr}");
+    gsub(/,[ ]?NULL[ ]?\)/, ", nullptr)");
+    gsub("    NULL)", "    nullptr)");
+    # special case: ", NULL, NULL,"
+    gsub(/,[ ]?NULL[ ]?,/, ", nullptr,");
+  }
```

Comment:
> Most (possibly all) of those cases should be removed using a script. I still support removing them that way!

Um. Lets call that script `scripts/maintenance/upgrade-NULL`.

I'm not sure why you call this:
 - "very complex" - it is literally a list of "replace X with Y" strings. regex only used to match optional whitespace. It cannot be simpler.
 - "unpolished" - it is polished per the style I use for awk scripts.
 - "dangerously unaware of C++ syntax" - not relevant, that is an issue for the author of careful "replace X with Y" pattern design (which has taken place).
 - "need has not been demonstrated" - need has been stated in your own "FWIW" paragraph. Demonstration is the number of changes it performs on current code.

Reply:
> > Most (possibly all) of those cases should be removed using a script. I still support removing them that way!
 
> Um. Lets call that script `scripts/maintenance/upgrade-NULL`.

I am against adding this script to the official repository, under any name.
 
> I'm not sure why you call this:
 
> * "very complex" - it is literally a list of "replace X with Y" strings. regex only used to match optional whitespace. It cannot be simpler.

The exact _effects_ of these regexes on C++ source code are very difficult (for a human) to predict correctly. I also disagree that all of these regexes are simple in isolation (some of them are not even readable IMO), but their individual complexity is a secondary problem.

The "it cannot be simpler" assertion, even if it is true, does not imply that the script is simple.

> * "unpolished" - it is polished per the style I use for awk scripts.

I said "unpolished", not "does not use style". If we wanted to accept this script, we would have polished it. That is extra work that we do not need to do if the script is not officially committed.

> * "dangerously unaware of C++ syntax" - not relevant, that is an issue for the author of careful "replace X with Y" pattern design (which has taken place).

It is relevant because these expressions, regardless of the author intent or convictions, can introduce unwanted changes to C++ code they are applied to. No review of past script applications can protect us from future application dangers (because the script does not understand even the basic structure of the source code it is modifying).

> * "need has not been demonstrated" - need has been stated in your own "FWIW" paragraph. Demonstration is the number of changes it performs on current code.

I am writing about the need for this script to be officially committed, not about the need for this (or some other) script to be used to replace some of the remaining NULLs in the official code.

Reply:
> The exact _effects_ of these regexes on C++ source code are very difficult (for a human) to predict correctly

Possibly, but they are easy - if tedious - to verify.
Does anyone have an actual example of any nefarious effects on the code by running this script now?
If there is none, given that the past is generally the best predictor of the future, there is little chance that there would be any. And even if, there are safeguards against that, both manual and automatic.

I wish to be practical, and apply this as a test: if we can find any actual examples where running the script causes problems, then it doesn't go in. But if we don't, then I support merging it, without any need for further polishing

Reply:
> Does anyone have an actual example of any nefarious effects on the code by running this script now?

I can come up with an example of code that this script screws up. I am sure anybody can do that if they wanted to. I assume that is not what you are asking for though.

Are you asking for an example of the _current_ code that this script screws up? Hopefully, such examples no longer exist because, during this script evolution, they were all fixed.

Or are you asking for an example of _likely_ future code that this script screws up? And after rejecting all examples of future code as "unlikely", will you then reject addition of this script itself because the future is unlikely to have a lot of new NULLs that will slip through, especially if we add automated checks that they do not?

> If there is none, given that the past is generally the best predictor of the future, there is little chance that there would be any. And even if, there are safeguards against that, both manual and automatic.

The bar for admission of problematic code (especially code-modifying code!) into the Squid repository is (or should be) much higher than "it does no harm to the current code (that we know about) and there is little chance it will do harm to future code".

The "future is like past" logic does not support the idea of adding this script either: After this script is merged, it will do _nothing_ to the then-current code. If the past is the best predictor of this script future, then, from that future point of view, we should assume that the merged script will continue to do _nothing_. We should not add scripts that are expected to do nothing.

> I wish to be practical, and apply this as a test: if we can find any actual examples where running the script causes problems, then it doesn't go in. But if we don't, then I support merging it, without any need for further polishing

I wish to be even more practical: Apply this script (if the author is comfortable with its results) but do not merge it, without any need for further searching, polishing, and without adding code corruption dangers and other problems. If we keep finding new NULLs being added to C++ code in master, then we can add an automated check that they are not added.

Reply:
Nothing short of a full AST parser would catch all theoretically possible cases, of course. And possibly not even there.

No, I mean is there anything wrong that is added to the actual code we use now. If there are none, there is no reason to not start mergin at least the results.

> I wish to be even more practical: Apply this script (if the author is comfortable with its results) but do not merge it, without any need for further searching, polishing, and without adding code corruption dangers and other problems. If we keep finding new NULLs being added to C++ code in master, then we can add an automated check that they are not added.

Here is where I disagree, and propose a more phased in approach:
- let's start merging the output now, and ensure it is stable
- then merge the script, and use it as part of source-maintenance
- if in the future it were to cause issues, then these will be caught by the test or review processes, and the author of the triggering patch to improve the script

what do you think?

Reply:
> is there anything wrong that is added to the actual code we use now.

I am not aware of any bugs created by this script, but I did not check carefully. I do not object to the results of this script being merged, under the (default) assumption that the author has checked those results carefully.

> there is no reason to not start mergin at least the results.

I agree! In fact, I have explicitly [supported](https://github.com/squid-cache/squid/pull/598#discussion_r1068708306) merging the results for a long time.

> I disagree, and propose a more phased in approach:

> * let's start merging the output now

We agree on the first step in your proposal. I still recommend making that step.

My objections regarding merging this script (i.e. the second proposed step) have not been addressed. FWIW, simply proposing steps that clearly go against my objections can rarely address my objections (but usually wastes time and nerve cells). However, there is no need to reach an agreement on that to make progress (i.e. to merge the script results vetted by the author).

Reply:
PR #1402 contains only the changes against latest master branch